### PR TITLE
Validate all generated meshes

### PR DESCRIPTION
**Pull request by pentacular**
The incremental processing of defective geometry is causing too many problems for the short-term speed increases.

Adds fallback repair paths from manifold to cgal.
Adds simple vertex splitting and separation to repair corner touches.
Disables manifold by default.

### CODE REVIEW

File: algorithm/cgal/repair_util.h [hunk 1]
```diff
@@ -157,3 +181,98 @@ static bool repair_self_intersections(Surface_mesh& mesh) {
...
+  return CGAL::Polygon_mesh_processing::remove_degenerate_edges(mesh) &&
+         CGAL::Polygon_mesh_processing::remove_degenerate_faces(mesh);
+}
+
+template <typename K>
+static bool repair_manifold(CGAL::Surface_mesh<typename K::Point_3>& mesh) {
+  typedef CGAL::Surface_mesh<typename K::Point_3> Surface_mesh;
+  // This repairs cases where we have two corners that touch at one coordinate.
+
+  // Duplicating the vertices fixes the topological problem, but not the
```
Comment (BarbourSmith) [inline algorithm/cgal/repair_util.h:196]: There are a lot of corner cases to handle 😉

**Review by BarbourSmith — APPROVED** [review]
Looks awesome!

If we convert say a square to manifold and then back to exact do we expect the square to be exactly the same? Is that a concern?


Do these look right?

image

**Comment by pentacular** [comment]
> Looks awesome!
> 
> If we convert say a square to manifold and then back to exact do we expect the square to be exactly the same? Is that a concern?
> 
> Do these look right?
> 
> image

Ah, I disabled the tests using Loft since it they were producing defective geometry.

Speaking of loft, I added an interesting paper "Optimal Boundary Triangulations of an Interpolating Ruled Surface" which seems to show how to do what I want much better. :)

I'll give fixing that a go in a bit.

File: algorithm/cgal/repair_util.h (hunk 1 above)
Comment (pentacular) [reply algorithm/cgal/repair_util.h:196]: Yeah, hopefully some are redundant or can be merged.

File: algorithm/cgal/boolean_util.h [hunk 2]
```diff
@@ -1,65 +1,110 @@
...
-    Surface_mesh meshCopy(b);
-    CGAL::Polygon_mesh_processing::reverse_face_orientations(
-        meshCopy);
-    if (!CGAL::Polygon_mesh_processing::clip(
-            a, meshCopy,
-            CGAL::parameters::use_compact_clipper(true),
-            CGAL::parameters::use_compact_clipper(true))) {
-      return false;
+/*
+The challenge with manifold is that conversion from exact to float can cause
```
[the remark this answers is not in the document]
Comment (BarbourSmith) [reply algorithm/cgal/boolean_util.h:15]: What if there are multiple shapes which share a face? After going through this process it's not guaranteed that they will share the face anymore, right?

[the remark this answers is not in the document]
Comment (pentacular) [reply algorithm/cgal/boolean_util.h:15]: Yes, there may now be a tiny crack or overlap.

I hope once the geometry is pretty solid, we can look into re-enabling manifold.

There are some techniques like negotiating common vertices between shapes that might make it work.

**Pull request merged**